compiler/src/parsing/parser.dyp

```diff
@@ -507,9 +507,13 @@ match_branches :
 match_expr :
   | MATCH lparen expr rparen lbrace match_branches rbrace { Exp.match_ ~loc:(symbol_rloc dyp) $3 $6 }
 
+list_expr_ending :
+  | comma? { None }
+  | comma ELLIPSIS expr { Some $3 }
+
 list_expr :
   | lbrack rbrack { Exp.list ~loc:(symbol_rloc dyp) [] None }
-  | lbrack expr [comma expr {$2}]* [comma ELLIPSIS expr {$3}]? rbrack { Exp.list ~loc:(symbol_rloc dyp) ($2::$3) $4 }
+  | lbrack expr [comma expr {$2}]* list_expr_ending rbrack { Exp.list ~loc:(symbol_rloc dyp) ($2::$3) $4 }
```

Comment:
@ospencer is there a way to improve this? I had to hoist the Some/None to the new rule because comma's value is unit.

Reply:
This is probably fine, though I might rename `list_expr_ending` to something like `list_spread_opt` or `list_spread`.

Reply:
I don't like naming this anything with `spread`, I actually considered the naming for a long time and it's not actually just a spread pattern but instead how the list expression "ends". Ideally, `comma ELLIPSIS expr` will be extracted into a `list_spread` rule when they are supported in more positions.